Firestore/core/src/firebase/firestore/objc/objc_class.h

```diff
@@ -0,0 +1,150 @@
+// headers that reference Objective-C types as members of C++ classes that must
+// be usable from straight C++.
+//
+// Care must be taken to not use the forward declaration (even implicitly) in
+// any inline definitions in the header. Even though Objective-C object
+// pointers look like raw pointers, under ARC they're more like
+// std::shared_ptr, where assignments and copies all potentially change the
+// refcount of the pointee. When methods that affect the reference count are
+// compiled inline in regular C++ this additional behavior won't get compiled
+// in and the Objective-C reference counts will be off.
+//
+// Note that this may even appear to work, though it does so through undefined
+// behavior: inline definitions are deduplicated at link time, and if the
+// linker happens to choose a definition that was generated in an ARC-enabled
+// translation unit then that specific build will work.
+//
+// Concretely this means that any method manipulating an Objective-C object
+// pointer (constructors, destructors, getters, and setters) all must be
+// defined out of line to avoid problems where ARC does not see changes to the
+// reference.
+//
+// To make this safe, any members of a type declared with OBJC_CLASS should be
+// held in a Handle object, defined below.
+#if __OBJC__
+#define OBJC_CLASS(name) @class name
+
+#else
+// Objective-C classes correspond to a C struct of the same name. Use a forward
+// declaration of that struct as the pointer type to help C++ generate the
+// correct mangled names.
+//
+// An alternative implementation would be to make this an alias for objc_object.
+// While that's appealing because it makes interaction with functions taking id
+// easier, any parameters of that type will be mangled differently in C++ vs
+// Objective-C++ leading undefined symbol errors at link time.
+#define OBJC_CLASS(name) struct name
+#endif  // __OBJC__
+
+/**
+ * A base class for implementing Handle, below, in such a way that allows all
+ * the major operations to be defined out-of-line in an ARC-managed translation
+ * unit.
+ */
+class HandleBase {
+ public:
+  HandleBase();
+
+  explicit HandleBase(id object);
+
+  ~HandleBase();
+
+  void Assign(id object);
+
+  void Release();
+
+ protected:
+  id object_;
+};
+
+/**
+ * A holder for a pointer to an Objective-C class, that can be declared in
+ * headers meant to be shared with plain C++. All useful operations on the
+ * handle are only available in Objective-C++.
+ */
+template <typename ObjcType>
+class Handle : public HandleBase {
+ public:
+  Handle() = default;
+
+  explicit Handle(ObjcType* object) : HandleBase(object) {
+  }
+
+#if __OBJC__
+  operator ObjcType*() const {
+    return static_cast<ObjcType*>(object_);
+  }
+
+  friend bool operator==(const Handle& lhs, const Handle& rhs) {
+    return (lhs.object_ == nil && rhs.object_ == nil) ||
+           [lhs.object_ isEqual:rhs.object_];
```

Comment:
Very optional: this is exactly what `objc::Equals` from `objc_compatibility.h` does. However, it may be the case that depending on that file is more trouble than its worth.